evals/registry.py

```diff
@@ -229,63 +229,73 @@ def _process_file(self, registry: RawRegistry, path: Path) -> None:
             return
 
         for name, spec in d.items():
-            assert name not in registry, f"duplicate entry: {name} from {path}"
-            if isinstance(spec, dict):
-                if "key" in spec:
-                    raise ValueError(
-                        f"key is a reserved keyword, but was used in {name} from {path}"
-                    )
-                if "group" in spec:
-                    raise ValueError(
-                        f"group is a reserved keyword, but was used in {name} from {path}"
-                    )
-                if "cls" in spec:
-                    raise ValueError(
-                        f"cls is a reserved keyword, but was used in {name} from {path}"
-                    )
+            yield name, path, spec
+
+    def _load_directory(self, path: Path) -> Generator[Tuple[str, Path, dict], None, None]:
+        files = Path(path).glob("*.yaml")
+        for file in files:
+            yield from self._load_file(file)
+
+    def _load_resources(
+        self, registry_path: Path, resource_type: str
+    ) -> Generator[Tuple[str, Path, dict], None, None]:
+        path = registry_path / resource_type
+        logging.info(f"Loading registry from {path}")
+
+        if os.path.exists(path):
+            if os.path.isdir(path):
+                yield from self._load_directory(path)
+            else:
+                yield from self._load_file(path)
+
+    @staticmethod
+    def _validate_reserved_keywords(spec: dict, name: str, path: Path) -> None:
+        for reserved_keyword in SPEC_RESERVED_KEYWORDS:
+            if reserved_keyword in spec:
+                raise ValueError(
+                    f"{reserved_keyword} is a reserved keyword, but was used in {name} from {path}"
+                )
+
+    def _load_registry(self, registry_paths: Sequence[Path], resource_type: str) -> RawRegistry:
+        """Load registry from a list of regstry paths and a specific resource type
+
+        Each path includes yaml files which are a dictionary of name -> spec.
+        """
+
+        registry: RawRegistry = {}
+
+        for registry_path in registry_paths:
+            for name, path, spec in self._load_resources(registry_path, resource_type):
+                assert name not in registry, f"duplicate entry: {name} from {path}"
+                self._validate_reserved_keywords(spec, name, path)
 
                 spec["key"] = name
                 spec["group"] = str(os.path.basename(path).split(".")[0])
+                spec["registry_path"] = registry_path
```

Comment:
reorganized spec loading to avoid weaving the registry_path down through functions just to add to the spec